sequences/seqs.gobra

```diff
@@ -0,0 +1,458 @@
+// non-empty, will be the last element of the latter sequence.
+ghost
+requires len(ys) > 0
+ensures Last(xs ++ ys) == Last(ys)
+decreases
+pure func LastOfConcat(xs, ys seq[int]) utils.Unit {
+	return utils.Unit{}
+}
+
+// Concatenating two sequences and then dropping the first element is equivalent
+// to dropping the first element of the first sequence and then concatenating
+// the sequences.
+ghost
+requires len(xs) > 0
+ensures DropFirst(xs ++ ys) == DropFirst(xs) ++ ys
+decreases
+pure func ConcatDropFirst(xs, ys seq[int]) utils.Unit {
+	return utils.Unit{}
+}
+
+// Concatenating two sequences and then dropping the last element is equivalent
+// to dropping the last element of the second sequence and then concatenating
+// the sequences.
+ghost
+requires len(ys) > 0
+ensures DropLast(xs ++ ys) == xs ++ DropLast(ys)
+decreases
+pure func ConcatDropLast(xs, ys seq[int]) utils.Unit {
+	return utils.Unit{}
+}
+
+// Returns true if there are no duplicate values in the sequence.
+ghost
+decreases
+pure func HasNoDuplicates(xs seq[int]) bool {
+	return forall i, j int :: { xs[i], xs[j] } (0 <= i && i < j && j < len(xs)) ==> xs[i] != xs[j]
+}
+
+// If a sequence does not contain duplicates, dropping the first or last value
+// yields a sequence that does not contain the dropped value.
+ghost
+requires len(xs) > 0
+requires HasNoDuplicates(xs)
+ensures !(First(xs) in DropFirst(xs))
+ensures !(Last(xs) in DropLast(xs))
+decreases
+// INFO Not too happy with the name.
+pure func DropExcludesValue(xs seq[int]) utils.Unit {
+	return utils.Unit{}
+}
+
+// For an element that occurs at least once in a sequence, the index of its
+// first occurrence is returned.
+ghost
+requires e in xs
+ensures 0 <= result && result < len(xs)
+ensures xs[result] == e
+ensures forall j int :: { xs[j] } (0 <= j && j < result) ==> xs[j] != e
+decreases xs
+pure func IndexOf(xs seq[int], e int) (result int) {
+	return First(xs) == e ? 0 : 1 + IndexOf(xs[1:], e)
+}
+
+// For an element that occurs at least once in a sequence, the index of its last
+// occurrence is returned.
+ghost
+requires e in xs
+ensures 0 <= result && result < len(xs)
+ensures xs[result] == e
+ensures forall j int :: { xs[j] } (result < j && j < len(xs)) ==> xs[j] != e
+decreases xs
+pure func LastIndexOf(xs seq[int], e int) (result int) {
+	return Last(xs) == e ? len(xs) - 1 : LastIndexOf(DropLast(xs), e)
+}
+
+// Returns a sequence without the element at a given position.
+ghost
+requires 0 <= pos && pos < len(xs)
+ensures len(result) == len(xs) - 1
+ensures forall i int :: { result[i] } { xs[i] } (0 <= i && i < pos) ==> result[i] == xs[i]
+ensures forall i int :: { result[i] } (pos <= i && i < len(xs) - 1) ==> result[i] == xs[i+1]
+decreases
+pure func Remove(xs seq[int], pos int) (result seq[int]) {
+	return xs[:pos] ++ xs[pos+1:]
+}
+
+// If a given element occurs at least once in a sequence, the sequence without
+// its first occurrence is returned.
```

Comment:
I feel like most of this descriptions are a contrived way of expressing what the functions are doing. Maybe we could simplify most of them and drop those that are really obvious.
```suggestion
// Deletes the first occurrence of e in xs.
```